website/blog/2019-07-15-complete-table-scan.md

```diff
@@ -0,0 +1,169 @@
+
+```sql
+SELECT SUM(l_extendedprice)
+FROM lineitem
+WHERE suppkey = 12345;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) |
+| -------- | ------------------- | ------------------ |
+| Aria     | 2                   | 14.2               |
+| Baseline | 3                   | 29.0               |
+
+
+This is a 2x win for the most basic case. This selects one row of a million and spends all its time in scan. It is here worthwhile to step through exactly how each implementation does this, as we need to understand this when analyzing the more complex cases.
+
+## Mechanics of a scan
+
+Baseline Presto does this as follows: The scan `OrcPageSource` produces consecutive `Page` instances that contain a `LazyBlock` for each column. This operation as such takes no time since the `LazyBlock` instances are just promises. The actual work takes place when evaluating the generated code for the comparison. This sees that the column is not loaded, loads all the values in the range of the `LazyBlock`, typically 1024 values and then does the operation and produces a set of passing row numbers. This set is empty for all but 1/1000 of the cases. If this is empty, the `LazyBlock` for `l_extendedprice` is not touched. If there are hits, the `l_extendedprice` `LazyBlock` is loaded and the values for the selected rows are copied out. When this happens, 1024 values are decoded from the column and most often one of them is accessed. Loading a `LazyBlock` allocates memory for each value. In the present case this becomes garbage immediately after first use. The same applies to the values in extended price, of which only one is copied to a `Block` of output. This is handled by a special buffering stage that accumulates rows from multiple loaded `LazyBlock` instances until there is a minimum batch worth of rows to pass to the next operator.
+
+The Aria implementation works as follows: The `OrcRecordReader` has an adaptable column order. It reads up to a full row group worth of data for the first column. When decoding the encoded values, it applies a callback to each value. The callback will either test the value and record the row number for passing values or copy out the value into a results buffer of the column or both. In the case of `suppkey = 12345` we only record the row number. Since for 99/100 row groups there is no hit we do not touch the second column. If there is a hit, we touch the hit rows of the second column. This has the effect of copying the decoded value into the column's result buffer. We keep going from row group to row group until enough data has been accumulated, at which point we return it as a `Page`. The `Block` instances in the `Page` are loaded and may either be views into the column reader's buffers or copies of the data, depending on what kind of downstream operator we have. If the downstream operator does not keep references to memory of processed `Block` instances, we can reuse the memory, which is owned by the column readers.
```

Comment:
`Since for 99/100 row groups` - would be nice to make this match `1/1000` above; otherwise it comes as a surprise

Reply:
Resolved by saying "virtualy all". There is such a thing as too precise.